File: tools/clouds_compare.cc
// Reference renderer for the `mi-clouds` port: dumps `<out_dir>/<MM>.pcm`
// (raw little-endian interleaved stereo int16) for every (playback mode,
// quality) pair, driven by a fixed deterministic parameter sweep over an
// arithmetic sawtooth input. The Rust `--example compare` produces the same
// files; `tools/wav_diff.py` reports the maximum per-sample delta.
//
// Clouds is a floating-point port with no bit-exactness contract, so a small
// non-zero delta is expected; it should stay within a few LSB.
//
// Build (from the eurorack C repo root, submodules checked out):
//   g++ -O2 -DTEST -I. -Istmlib -o /tmp/clouds_compare \
//       ../eurorack-rs/tools/clouds_compare.cc \
//       clouds/dsp/granular_processor.cc clouds/dsp/correlator.cc \
//       clouds/dsp/mu_law.cc clouds/dsp/pvoc/phase_vocoder.cc \
//       clouds/dsp/pvoc/stft.cc clouds/dsp/pvoc/frame_transformation.cc \
//       clouds/resources.cc stmlib/utils/random.cc stmlib/dsp/units.cc \
//       stmlib/dsp/atan.cc
//   mkdir -p /tmp/c_pcm && /tmp/clouds_compare /tmp/c_pcm

#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <new>

#include "clouds/dsp/granular_processor.h"
#include "stmlib/utils/random.h"

using namespace clouds;
using stmlib::Random;

static const int kBlock = 32;
static const int kBlocks = 1500;
// The firmware runs Prepare() in a tight while(1) loop between audio blocks;
// the WSOLA correlator search is spread across those calls.
static const int kPrepareIters = 32;

static const PlaybackMode kModes[3] = {
    PLAYBACK_MODE_GRANULAR,
    PLAYBACK_MODE_STRETCH,
    PLAYBACK_MODE_LOOPING_DELAY,
};

static float tri(float x) {
  x = fabsf(x - floorf(x));
  return x < 0.5f ? x * 2.0f : 2.0f - x * 2.0f;
}

// Same sizes as clouds.cc (block_mem / block_ccm).
static uint8_t large_buffer[118784];
static uint8_t small_buffer[65536 - 128];

int main(int argc, char** argv) {
  const char* out_dir = argc > 1 ? argv[1] : ".";

  for (int mode_idx = 0; mode_idx < 3; ++mode_idx) {
    for (int quality = 0; quality < 4; ++quality) {
      Random::Seed(0x21);

      void* mem = calloc(1, sizeof(GranularProcessor));
      GranularProcessor& gp = *new (mem) GranularProcessor();
      memset(mem, 0, sizeof(GranularProcessor));
      memset(large_buffer, 0, sizeof(large_buffer));
      memset(small_buffer, 0, sizeof(small_buffer));

      gp.Init(large_buffer, sizeof(large_buffer),
              small_buffer, sizeof(small_buffer));
      gp.set_playback_mode(kModes[mode_idx]);
      gp.set_quality(quality);
      for (int k = 0; k < kPrepareIters; ++k) gp.Prepare();

      char path[512];
      snprintf(path, sizeof(path), "%s/%02d.pcm", out_dir, mode_idx * 4 + quality);
      FILE* fp = fopen(path, "wb");

      float phase = 0.0f;
      for (int i = 0; i < kBlocks; ++i) {
        float t = (float) i / (float) kBlocks;
        Parameters* p = gp.mutable_parameters();
        p->position = tri(t * 2.0f);
        p->size = 0.2f + 0.6f * tri(t * 3.0f + 0.1f);
        p->pitch = -7.0f + 14.0f * tri(t * 1.3f);
        p->density = 0.3f + 0.5f * tri(t * 2.5f + 0.3f);
        p->texture = tri(t * 1.7f + 0.5f);
        p->dry_wet = 1.0f;
        p->stereo_spread = 0.5f;
        p->feedback = 0.0f;
        p->reverb = 0.0f;
        p->freeze = (i / 200) % 3 == 2;
        p->trigger = i % 48 == 0;
        p->gate = false;

        ShortFrame input[kBlock];
        ShortFrame output[kBlock];
        for (int j = 0; j < kBlock; ++j) {
          phase += 220.0f / 32000.0f;
          if (phase >= 1.0f) phase -= 1.0f;
          int16_t s = (int16_t) ((phase - 0.5f) * 24000.0f);
          input[j].l = s;
          input[j].r = s;
        }

        gp.Process(input, output, kBlock);
        for (int k = 0; k < kPrepareIters; ++k) gp.Prepare();
        fwrite(output, sizeof(ShortFrame), kBlock, fp);
      }
      fclose(fp);
      free(mem);
    }
  }
  return 0;
}

File: crates/clouds/examples/clouds_compare.rs
//! Rust side of the C<->Rust comparison. Mirrors `tools/clouds_compare.cc`
//! exactly: for every (playback mode, quality) pair it renders the same
//! deterministic parameter sweep over an arithmetic sawtooth input and dumps
//! `<out_dir>/<MM>.pcm` (raw little-endian interleaved stereo i16).
//!
//! Clouds is a floating-point port with no bit-exactness contract, but 10 of
//! the 12 dumps still come out byte-identical; `tools/wav_diff.py` reports the
//! maximum per-sample delta for the rest.
//!
//!   cargo run --release --example clouds_compare -p mi-clouds -- /tmp/rust_pcm
//!   # then build + run tools/clouds_compare.cc into /tmp/c_pcm and:
//!   python3 tools/wav_diff.py /tmp/c_pcm /tmp/rust_pcm

use std::env;
use std::fs;
use std::io::Write;

use clouds::{GranularProcessor, PlaybackMode, ShortFrame};
use stmlib::Random;

// A plain arithmetic sawtooth -- broadband test material with no
// transcendental function, so the C and Rust inputs are bit-identical and
// any output delta is the DSP port's, not `sinf`'s.

const BLOCK: usize = 32;
const BLOCKS: usize = 1500;
/// The firmware runs `Prepare()` in a tight `while (1)` loop between audio
/// blocks; the WSOLA correlator search is spread across those calls. Mirror
/// that with a fixed budget per block so Stretch mode actually produces
/// sound.
const PREPARE_ITERS: usize = 32;

const MODES: [PlaybackMode; 3] = [
    PlaybackMode::Granular,
    PlaybackMode::Stretch,
    PlaybackMode::LoopingDelay,
];

fn tri(x: f32) -> f32 {
    let x = (x - x.floor()).abs();
    if x < 0.5 {
        x * 2.0
    } else {
        2.0 - x * 2.0
    }
}

fn main() {
    let out_dir = env::args().nth(1).unwrap_or_else(|| ".".to_string());
    fs::create_dir_all(&out_dir).unwrap();

    for (mode_idx, mode) in MODES.into_iter().enumerate() {
        for quality in 0..4i32 {
            Random::seed(0x21);
            let mut gp = GranularProcessor::new();
            gp.set_playback_mode(mode);
            gp.set_quality(quality);
            for _ in 0..PREPARE_ITERS {
                gp.prepare();
            }

            let mut phase = 0.0f32;
            let mut bytes = Vec::with_capacity(BLOCKS * BLOCK * 4);
            for i in 0..BLOCKS {
                let t = i as f32 / BLOCKS as f32;
                {
                    let p = gp.mutable_parameters();
                    p.position = tri(t * 2.0);
                    p.size = 0.2 + 0.6 * tri(t * 3.0 + 0.1);
                    p.pitch = -7.0 + 14.0 * tri(t * 1.3);
                    p.density = 0.3 + 0.5 * tri(t * 2.5 + 0.3);
                    p.texture = tri(t * 1.7 + 0.5);
                    p.dry_wet = 1.0;
                    p.stereo_spread = 0.5;
                    p.feedback = 0.0;
                    p.reverb = 0.0;
                    p.freeze = (i / 200) % 3 == 2;
                    p.trigger = i % 48 == 0;
                    p.gate = false;
                }

                let mut input = [ShortFrame::default(); BLOCK];
                for frame in input.iter_mut() {
                    phase += 220.0 / 32000.0;
                    if phase >= 1.0 {
                        phase -= 1.0;
                    }
                    let s = ((phase - 0.5) * 24000.0) as i16;
                    frame.l = s;
                    frame.r = s;
                }

                let mut output = [ShortFrame::default(); BLOCK];
                gp.process(&input, &mut output);
                for _ in 0..PREPARE_ITERS {
                    gp.prepare();
                }

                for frame in output {
                    bytes.extend_from_slice(&frame.l.to_le_bytes());
                    bytes.extend_from_slice(&frame.r.to_le_bytes());
                }
            }

            let idx = mode_idx * 4 + quality as usize;
            let path = format!("{out_dir}/{idx:02}.pcm");
            fs::File::create(&path).unwrap().write_all(&bytes).unwrap();
        }
    }
    println!("wrote {} dumps to {out_dir}", MODES.len() * 4);
}
